### ISSUE TITLE
[TW#14142] Bluetooth Mesh support (IDFGH-763)

### ISSUE THREAD

**Issue by OtherCrashOverride** [espressif/esp-idf#819]
Support for the new Bluetooth Mesh standard needs to be addressed.

**Comment by FHFS** [comment]
Do you mean this announcement by [SIG](https://www.bluetooth.com/news/pressreleases/2017/07/bluetooth-sig-announces-mesh-networking-capability) not more than 24 hours ago?

**Comment by OtherCrashOverride** [comment]
Yes.  Nordic and others have a "day one" SDK available.

**Comment by FHFS** [comment]
Does sound like a nice feature =)

**Comment by TianHao-Espressif** [comment]
Thanks for your attention to ESP32. Standard Mesh Support is in plan, but the time need to discuss.

**Comment by sthenc** [comment]
Hello, I am also interested in this. Any updates on the status?

Btw. the specification can be accessed publicly, I just downloaded it:
https://www.bluetooth.com/specifications/mesh-specifications

I fail to understand how big a difference there is between BLE mesh and normal BLE (but my understanding of both is very limited). 

I have a requirement for two BLE slave devices to exchange data between themselves.

@TianHao-Espressif  Do you think it would be possible to build a BLE mesh on top of the present esp32-bt-lib?
If implementing BLE mesh for ESP32 is not going to happen soon (meaning a couple of weeks), are there some already working alternatives?

**Comment by FHFS** [comment]
@sthenc I think GATT server-client model would do for you.
[Mesh](https://en.wikipedia.org/wiki/Mesh_networking) is a network type, where you would be able to communicate with all devices on that network, even when for example that other node would be out of range.

**Comment by jrewiens** [comment]
Is there any progress on ESP32 BLE mesh functionality?

**Comment by TianHao-Espressif** [comment]
@jrewiens , for now, BLE -MESH is not ready in ESP32. You can consider about ESP-MESH with WIFI.

**Comment by micheljung** [comment]
ESP-Mesh isn't ready for ESP32 either.

I'm willing to pay for BLE-Mesh support.

**Comment by benjaminaigner** [comment]
@micheljung 
There is the BTSTack BLE stack, where they are currently developing BLE Mesh support.
I think @mringwal will be faster than the Bluedroid stack :-):

https://github.com/bluekitchen/btstack/tree/mesh

**Comment by mringwal** [comment]
Bluetooth Mesh works by sending and receiving advertisements. I've seen that Espressif added support for scanning and advertising at the same time (well, it stops quickly to receive when sending). That's the basis for any Mesh implementation.

@benjaminaigner thanks for your confidence. We're fuzzy about the when, but we're working definitely on Bluetooth Mesh for BTstack (a few PTS tests are already green). It does have a few protocol layers to implement.

**Comment by nvl1109** [comment]
Hello @TianHao-Espressif  and all,
Do you have any updates about BLE mesh for ESP32? Thank you.

**Comment by Yulong-espressif** [comment]
@nvl1109 About the BLE Mesh. We are currently working on it and it will take time, so stay tuned for our follow-up IDF update. Thank you

**Comment by jo-nas** [comment]
Any News?

**Comment by mringwal** [comment]
BTstack's BLE Mesh supports Provisioning both as Device (new device that gets added to the network) and Provisioner (the device that provides network key to new device - similar to LE Secure Connections) in the test/mesh folder of the "mesh" branch. We plan to have basic network communication working maybe end of January.

**Comment by Yulong-espressif** [comment]
@jo-nas We will support BLE Mesh at the first quarter of 2018. We are writing the code now. Thanks for your concerned.

**Comment by NicksonYap** [comment]
@Yulong-espressif @TianHao-Espressif  
any news or beta release of BLE Mesh?

**Comment by Yulong-espressif** [comment]
@NicksonYap We have now implemented some of the features of Bluetooth mesh. We plan to release the first version of the Bluetooth mesh in early April. Thank you for your interest in the progress of Bluetooth mesh.

**Comment by wolfelec** [comment]
Hello Yulong-espressif -,

Will the bluetooth mesh be available with AT commands?

Thank you :)

**Comment by wolfelec** [comment]
Hello @Yulong-espressif 

Will the bluetooth mesh be available with AT commands?

Thank you :)

**Comment by Yulong-espressif** [comment]
@wolfelec We currently have no plans to join the Mesh AT command, we will consider other BLE Mesh release, thank you.

**Comment by wolfelec** [comment]
@Yulong-espressif 

Thank you for your answer :)

So if I understand correctly, the new functionality expected for April, will not bring AT specific command to BLE Mesh?

Thank you

**Comment by herry13** [comment]
Pin this thread because I'm very interested to use BLE mesh for my project.

**Comment by micheljung** [comment]
@Yulong-espressif how's it going, are we on track? Early April ends soon :-)

**Comment by rapzak** [comment]
@Yulong-espressif,
Hi - I am also interested in this :) :)

**Comment by chegewara** [comment]
Few more interested people on this and i will create account on kickstart and write it by myself 🥇

**Comment by sh4kesbeer** [comment]
Have been waiting to test BLE Mesh for two years (since the earliest rumors) now, would be great if you could bring it to the esp!

**Comment by Yulong-espressif** [comment]
@sh4kesbeer @rapzak @micheljung Sorry for the delay, The BLE mesh code already finish now, we still need some time to test it, we will release it as soon as possible to the IDF. Thank you very much.

**Comment by ZeeLivermorium** [comment]
Thank god! Y'all da best!

**Comment by wolfelec** [comment]
@Yulong-espressif , thank you for the new :)

**Comment by mik00** [comment]
Looking forward to BLE mesh release too.

**Comment by manikandan-enability** [comment]
Any Update about BLE mesh for ESP32?
Thank you. 
@Yulong-espressif

**Comment by leodesigner** [comment]
Maybe http://zephyrproject.org/ will be a faster solution for BLE Mesh ?

**Comment by micheljung** [comment]
Last time I checked, they didn't even have bluetooth drivers fro ESP32 yet. I paid $100 via Bountysource so someone implemented it, but IIRC there were many other shortcomings as well so I gave up on Zephyr. @leodesigner

**Comment by Shannon-I-am** [comment]
How is that testing going?
Can we expect BLE mesh to work with the existing wifi mesh interfaces?
I'm anxious to have this functionality.

**Comment by chegewara** [comment]
I see new bluetooth option, a lot bluetooth options, one is suggesting we are having or we are very close to have implemented ble mesh. Some confirmation? @Weijian-Espressif @Yulong-espressif

**Comment by micheljung** [comment]
@Yulong-espressif first you announced March, then you moved it to "early April", then you said you were "done" and now it's JUNE and we're still not there? I'm trying to run a business and I've been delaying my product waiting for BLE Mesh, if I knew these anouncements were so unreliable, I would've looked for another solution.

This really hurts Espressif's reputation. And my business.

**Comment by Yulong-espressif** [comment]
@micheljung I'm sorry about the delay,  we have now completed the development of the BLE Mesh code protocol part. We can only provide to some customers at present. If you are in urgent need, please contact our business department. We can give priority to you. Feel sorry again.

**Comment by Shannon-I-am** [comment]
@Yulong-espressif, will BLE Mesh ever be available without a 'special arrangement'? Maybe we should start looking elsewhere immediately?

**Comment by denyip** [comment]
Hi, Is it possible to make a ESP32 as a bluetooth 5 device provisioner?

**Comment by sh4kesbeer** [comment]
@OtherCrashOverride well, they have no obligation whatsoever to fulfil your demands, maybe you should be a bit more grateful for the fact that they are listening at all?

**Comment by NicksonYap** [comment]
Let's keep this simple.
As @sh4kesbeer said, they have no obligation
A GitHub comment is not an official release date
It's our own responsibility to keep a look out at alternatives
This is a low cost platform, there's nothing much to expect

**Comment by Yulong-espressif** [comment]
@all Thank you for your continuing interest in ESP32 BLE Mesh. At the moment we have decided internally that if there is no problem with the test, we will publish the BLE Mesh function to Github at the end of July.

**Comment by wolfelec** [comment]
@Yulong-espressif Good news :)

**Comment by OtherCrashOverride** [comment]
@sh4kesbeer 

> well, they have no obligation whatsoever to fulfil your demands

It was not a demand.  It was a request from a **customer**.  No threats were made.  No ultimatums were given (by me). Almost a year after my first inquiry, I have followed up on it.

> maybe you should be a bit more grateful for the fact that they are listening at all?

When a customer asks a question to a business, it is called **customer service**.  The level of customer service is also considered a very important measure of a company by many.  

As expressed earlier in the conversation, customers have their own businesses to run and their own schedules.  It is common practice in the industry to inquire about product "roadmaps".  If a vendor is not **reasonably** forthcoming about their "roadmaps" or chastises their customers for inquiring, then a customer may seek a relationship with a different vendor.  Its not an emotional reaction; rather, its simply common business practice in a highly competitive and time sensitive market.

Modern development is increasingly decentralized (global).  In this specific instance, it far simpler to select an existing BT Mesh solution than to work within the constraints of a 'special arrangement'.  When all development interaction has to go through 'approved channels', it significantly adds to the time and resources required for product development.

**Comment by denyip** [comment]
@Yulong-espressif , do you mean that the release on Jul will be compatible to Bluetooth 5 mesh, that means it can communicate with other brand chips via Bluetooth 5 mesh? If it's true, then will be great.

**Comment by Yulong-espressif** [comment]
@denyip ESP32 currently does not support the Bluetooth 5.0 feature, but BLE Mesh and 5.0 are independent. This means that our BLE Mesh is compatible with BLE Mesh on other chips, so feel free to use it. Thank you very much.

**Comment by H0bbes** [comment]
Thank you for your hard work. Can't wait to start playing with it.
Did you solved/optimized the cohabitation between BLE scan/advertising and Wifi ?

**Comment by denyip** [comment]
yes, it's basically, the system connect to wifi aws and then do scan ble simultaneously, run few months without problem.

**Comment by ramams** [comment]
@Yulong-espressif is ESP BLE Mesh released? From your previous message it was expected to be released by end of July. Thanks in advance.

**Comment by ngonhatnhi** [comment]
@Yulong-espressif Can you tell us released plan of ESP BLE Mesh ? We had to wait in very long time...

**Comment by Shannon-I-am** [comment]
@ramams and @ngonhatnhi, I guess there was a problem with the test.

**Comment by ngonhatnhi** [comment]
I will attempt to contact Espressif biz service, hope that we 'll get some news about BLE Mesh

**Comment by Yulong-espressif** [comment]
Hi, @all, Thank you for your continued attention to BLE Mesh. At present, ESP32's BLE Mesh has been opened to the second batch of customers for testing. Due to business considerations, we are not yet able to publicly release the BLE Mesh version. If you need it urgently, please contact our business department. I am sorry for the inconvenience caused to you.

**Comment by ngonhatnhi** [comment]
@Yulong-espressif Hello Mr.Yulong, i can not find email and phone number of your biz deparment in espressif.com , so can you help me ?

**Comment by Shannon-I-am** [comment]
With such lack of clarity and inability to keep to schedules I'm switching from Espressif to Nordic.

**Comment by jo-nas** [comment]
What about this? https://github.com/espressif/esp-mdf

**Comment by Shannon-I-am** [comment]
@jo-nas that uses wifi, so although it may be relatively low power it's not going to be able to provide the battery-life of a BLE Mesh network.

**Comment by ReanimationXP** [comment]
@Yulong-espressif any update on public release?

**Comment by toxuin** [comment]
I think we should bump this more often.

**Comment by sthenc** [comment]
This is taking so long I completely lost interest in this. 

With this not being done more than a year after the initial plans (and no one reverse-engineering the thing), I think it might even be possible they had some sort of silicon bug. Priority customers usually get the engineering samples first for the new silicon, and this is probably how long it takes to fix that sort of bug.

But I am not sure which is more disturbing, complete lack of transparency regarding that, or restricting access to basic functionality that the competitors have had for a year or so now.

**Comment by DevyatovAndrey** [comment]
@Yulong-espressif any update on public release?

**Comment by Isl2017** [comment]
Hi, All

Thanks for your interest in our ESP32 and BLE Mesh. 

About the BLE Mesh development
* In order to make the most of our limited human resource on the development of new “BLE Mesh Features” and focus on supporting some important customers, currently the ESP BLE Mesh SDK is not public and is provided only to some internal-test customers. 
* The latest version of ESP BLE Mesh is “ESP-BLE-Mesh-v0.5” and it has been just provided to our round-2 internal test customers.
* If you want to participate in our internal test, it is recommended to contact our Business team for more information
* So currently there is no exact public release plan. If your project is urgent, please participate in our internal-test

Note:  “ESP-BLE-Mesh-v0.5” supports the following features
1. Provisioning
        * Support PB-ADV and PB-GATT bearers
        * Add Provisioner feature
2. Networking
        * Support relay, segmentation, etc
3. Proxy Support
4. Foundation Models
        * Configuration Server Model
        * Configuration Client Model
        * Health Server Model
        * Health Client Model
5. Generic Client Models
        * Generic OnOff Client
        * Generic Level Client
        * Generic Default Transition Timer Client
        * Generic Power OnOff Client
        * Generic Power Level Client
        * Generic Battery Client
        * Generic Property Client
6. Lighting Client Models
        * Light Lightness Client
        * Light CTL Client
        * Light HSL Client
7. Sensor Client Models
        * Sensor Client
8. Time and Scenes Client Models
        * Scene Client

**Comment by chegewara** [comment]
Glad to hear that rest of us (read most of us) are not important customers.

**Comment by Isl2017** [comment]
> Glad to hear that rest of us (read most of us) are not important customers.

I'm sorry for my inappropriate description. But I hope you can understand us. As of right now, we have accepted all the customers who applied for the internal test. Please contact business team if you really need it currently, thanks :)

**Comment by chegewara** [comment]
@Isl2017 I dont need it urgently, probably i dont need it at all, but i would like to participate in test.
Its not like that i feel i am good at bluetooth, but as you said, you are having limited human resources.

**Comment by Isl2017** [comment]
@chegewara Really thanks a lot for your attention and support. We hope that our BLE Mesh is somewhat stable and can support most necessary features when we release it. Hope again for your understanding :)

**Comment by chegewara** [comment]
Thats the problem i dont understand. Thinking like that then whole esp-idf stack should be closed for only few testers, because there still is a whole bunch of issues/bugs. Dont get me wrong, i dont complain about that. This is big project and i know that a lot good programmers are working on it. When first version of esp-idf has been released how well developed it was?

What you said about being important customer, i understand that too. I am not important to you (espressif) because i wont buy 10, 20 and maybe even 100k chips/modules. But i would rather prefer to wait for ble mesh not knowing that. So your statement was very loud and clear. 

Have a nice day.

**Comment by Isl2017** [comment]
To apply for the internal test, please feel free to submit your demands here : https://www.espressif.com/en/company/contact/pre-sale-questions-crm

The submitted questions will be transferred to our CRM system directly, and the BS team will get back to you asap.

**Comment by ReanimationXP** [comment]
I agree that the entire point of open source is being missed here. Opening it up to others would allow them to assist you with pull requests. Nobody is going to bash you for a properly marked experimental beta not working correctly.. if anything they would try to help get it working and flushing out the feature set. You shouldn't need details on whatever project we're experimenting with or a personal consultation in order to share some framework code with us and let us determine if it's presently useful for us or not. It sounds to me like you're focusing too much on personal partnerships instead of getting the average developer comfortable with new features of your product. You can trust that the latter will go a lot further in future sales.

**Comment by lucascoxBAF** [comment]
@Isl2017 , I couldn't agree with @chegewara more. If there is a working, buggy ble mesh stack then please release it as an alpha stack and warn that it is not ready for production code. Open source developers can then provide input and potential bug fixes. As two others have said, you are missing the point of open source.

**Comment by chegewara** [comment]
@Isl2017 Little offtopic:
> @chegewara Really thanks a lot for your attention and support. We hope that our BLE Mesh is somewhat stable and can support most necessary features when we release it. Hope again for your understanding :)

To be honest i dont think that espressif team cares about my attention or support at all. But here is the thing, i dont feel like i have skills to prepare some PRs in esp-idf or even arduino-esp32 repository. I really suck at programming skills, but i have plenty of time at my regular work and i can do whatever i want, watch tv or play games or play with any other hardware. So why i am spending so much time with esp32? Because i think its very nice piece of hardware which can be used to build so many cool stuff. But thats not all, here on esp-idf repository issues i have contact with few people (i am opening a lot issues), and personally i think good about them. Thats why i think its worth to support esp32 and this project. For some reason support for ble and bluetooth in generally on forum is very poor, its main reason i am opening so many issues here.

But what i want to say. I dont want to start storm in the glass, i am just trying that you guys have support from community and we community prefer rather to have piece of software with basic functionality to learn it, to help you make it better etc rather than be treat like that. There was announcement that we get mesh in april, then july and so on, now we hear that we are not important customers and we get it when we get it. This personally makes me feeling that ive been wrong and i become feeling i want to quit this and just play with project that @nkolban start. This gives me fun and pleasure because all what i am doing someone else will shows me appreciation just by using piece of code i write.

**Comment by bwannduke** [comment]
I definitely agree with getting early-alpha out to the community, but there is a fine line with open-source, specifically with a commercial offering of hardware behind it.  If Espressif releases initial releases that are buggy as hell, and not even the slightest ready, it reflects poorly on them, and they are pushing a new model of open-source in an embedded space to cutthroat and historically very proprietary and litigious.  I think it is the most important message to send to them as the community of devs that want to work on this product platform, that their product is being adopted, and that the community of developers themselves have helped to establish ESP's brand in the market as an open platform, and that we are ready to dive in and help them release this product to the market.

**Comment by todduk** [comment]
Please top whining.

**Comment by bwannduke** [comment]
Going to pass on this flame bait

**Comment by ReanimationXP** [comment]
ESP32 already has an established foothold, nothing they release at this point is going to harm the product or the brand, so that excuse is crap. Besides that, they are already in the testing phase and have been for months, so they already have something that is at least somewhat functional. This is what alphas and betas are for. Just because you might not be comfortable with submitting PRs, that doesn't mean the rest of the world isn't. There's still plenty of stuff you could do such as documentation to help them out that they would be wasting time on internally because they didn't open source the project.

**Comment by denyip** [comment]
@Isl2017 , I have followed your instruction to submit the application form for internal testing, but there is no any reply at all.

**Comment by Isl2017** [comment]
@chegewara @ReanimationXP @bwannduke Thanks for all your suggestions. I keep listening to your comments and thinking over whether it is possible to release BLE Mesh SDK earlier. Please help understand that "BLE Mesh is not somewhat stable now" is not the only reason why we decided to release it later, but there are some other considerations.

For example, we need to observe our “IDF breaking change rules” once it is merged into the IDF. That means, once “BLE Mesh feature” is merged to the IDF,  we can’t add breaking changes (like API modification)  to the IDF except for a major IDF version (like IDF 3.0, 4.0, etc). It will limit our mesh development. Currently, I have been discussing with our IDF team to see if it is possible to add "BLE Mesh feature" as a beta version and not observe this "breaking change rule".

We will try our best to resolve those difficulties, release BLE Mesh SDK earlier and hope that you could understand our difficulties. Please help wait for some more time and we will publish an announcement here once we have any updates for release.

**Comment by Isl2017** [comment]
@denyip I have checked with our Business team and they have not found the requirement with keyword "denyip". Could you re-submit your requirement and leave your message title here? We will re-check it.

**Comment by pctj101** [comment]
Probably reasonable to put such unstable API code in a branch so that it's pretty obvious the API is under development.

**Comment by Isl2017** [comment]
@pctj101 Yes, it is another solution we are considering. The problem we encounter for this solution is:      
* Because we separate the features of IDF and the features of BLE Mesh;
* So for some features that we need to modify the IDF, we will merge it into the IDF firstly and then rebase our BLE Mesh branch (we will not introduce breaking changes when we add the features for IDF)
* In this case, if we use this solution (put the BLE Mesh feature in a branch), it might difficult for us to manage different versions
* Because once we rebase the BLE Mesh branch, the customer can’t revert to the older version

**Comment by pctj101** [comment]
Yeah sounds like to minimize pre-release work, a non-maintained branch may be more fitting. A branch which kinda works but is not scheduled to get any updates.  Perhaps even if tight on time, without full integration with the IDF. 

Any updates would go into another branch which might have no easy migration path.

It might not be convenient for early developers but sometimes thats the cost of being on the bleeding edge.  If you set the expectations for support/roadmap (for example “no support/not taking PRs yet” then there’s little basis for complaint). 

Each developer eventually needs to learn for themselves when they have sufficient skill to use a API in various stages of the lifecycle. If it’s too hard then wait. 

Anyways good to see your team is considering various options!

**Comment by ReanimationXP** [comment]
@Isl2017 Yes, thanks for listening and providing some rationale. It's understandable that Mesh requiring changes to the IDF which would break existing apps is not a great solution to push out before that version of the IDF is ready.

I'm on board with @pctj101's suggestion of a dead-end branch of the IDF which includes the necessary changes for Mesh, such that we could play with it in the same environment you develop on. I will add that it would be a good idea to suggest that users simply clone down a second copy of the "beta" IDF specifically for working with Mesh. Once you merge the changes into the next major IDF version, the user would simply update their main IDF as normal, and then projects written for mesh would likely transfer over with little to no issue.

**Comment by pctj101** [comment]
There might be a case where some devs decide to submit high numbers of issues/tickets for the BLE Mesh, even though it's been publicly stated that it's unsupported.  

In that case, it might be good to make a separate esp-blemesh-beta repo for example and just blackhole all the tickets (or read them later when ready to release officially).

**Comment by ReanimationXP** [comment]
Again I agree, getting issues / tickets for something that is clearly in beta would not be great, so a separate repo entirely might be wise. Additionally or alternatively, marking it as an alpha might discourage issue creation further. Our intent definitely isn't to hamper development with bogus tickets.

**Comment by chegewara** [comment]
It is possible that wifi mesh is in separate repository, esp-adf too, even if its in very early stage. Both are linked to esp-idf particular commit. I am listening all arguments and still dont see reason why it cant be put in separate repository if so many function is already implemented. 
I dont know if you remember what situation with esp-idf was about year ago or so. There was group of people complaining that esp-idf is so bugged that should not be released. But some of users, including me kept telling them it is staged as beta and its normal that we have bugs. esp-idf survived such complaining people but thanks to issues reported by community it was developed to stage we have today so much faster in my opinion. 

I understand that espressif decision is we wont have access to bt mesh yet and to be honest i dont need explanations and rational arguments. Frustrating is that since this issue has been opened we got few possible dates when we get it, then about half year without update and now this. "We have nice looking working beta version but it is for few chosen people only".

**Comment by Isl2017** [comment]
Hi, All 

Thanks for all your brainstorms.

We have created a temporary BLE Mesh branch on Github now.

It is now allowed to clone the BLE Mesh project:
* Branch: [feature/esp-ble-mesh-v0.5](https://github.com/espressif/esp-idf/tree/feature/esp-ble-mesh-v0.5)
* BLE Mesh APIs & Core code: [components/bt/ble_mesh](https://github.com/espressif/esp-idf/tree/feature/esp-ble-mesh-v0.5/components/bt/ble_mesh)
* BLE Mesh Examples:  [examples/bluetooth/ble_mesh](https://github.com/espressif/esp-idf/tree/feature/esp-ble-mesh-v0.5/examples/bluetooth/ble_mesh)

Please help understand that this is a beta version and I think there's still a lot to be improved.

We hope to receive your suggestions about the ESP BLE Mesh SDK. So if you have any issues/advice/questions/feature requests, don't hesitate to raise [issues](https://github.com/espressif/esp-idf/issues) on Github.

When you raise issues/questions/requests, please help add a tag of BLE Mesh to the issue title, so that we could quickly recognize and track. Thanks.

**Comment by pctj101** [comment]
Sweet!

**Comment by chegewara** [comment]
@Isl2017 Thanks you for very good tutorials. As always nice job.

**Comment by Isl2017** [comment]
@pctj101 @chegewara Thanks for your approvals. We are going into it 🙂

**Comment by chegewara** [comment]
@Isl2017 I was harsh, but that does not mean i dont appreciate hard and good work you all are doing.
Ive seen all ble tutorials and they are very helpful. When i can i paste link to it. Its always nicely write, easy to understand and all info needed to understand API usage.

**Comment by chegewara** [comment]
I have to say BIG thank you for releasing beta version. I had not much time to play with it so far, but i love it. I dont know yet which android provisioner i like more, with silicon labs we have options to easy switch all available server configs like proxy, relay or element type. In other hand is not recognizing more than one element per node (maybe its esp32 issue see ad1) even if is assigning all necessary unicast addresses. Second provisioner is nRF Mesh, which is not as easy to use, but is recognizing all elements in one mesh node (3 in my tests). I could not find option to turn on/off relay or proxy like in Silicon lab app, but there is more options to manage mesh node.

AD1: I dont want to open issue yet, because i dont know if this is issue or just i am initializing elements in node wrong way. This is log i am getting:
```
E (448) BT_LOG: btc_mesh_model_op_add, Invalid model operation min_len = 1073486928, or param
callbak = 0x3ffc1d48, the last valid opcode = 0x0
E (448) BT_LOG: btc_mesh_model_op_add, Invalid model operation min_len = 1073486928, or param
callbak = 0x3ffc1d48, the last valid opcode = 0x0
E (458) BT_LOG: btc_mesh_model_op_add, Invalid model operation min_len = 1073486928, or param
callbak = 0x3ffc1d48, the last valid opcode = 0x0
```
Code:
```
static esp_ble_mesh_model_t root_models[] = {
    ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
    ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op, &onoff_pub, &led_state[0]),
};

static esp_ble_mesh_model_t root_models1[] = {
    ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
    ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op, &onoff_pub, &led_state[1]),
};

static esp_ble_mesh_model_t root_models2[] = {
    ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
    ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op, &onoff_pub, &led_state[2]),
};

static esp_ble_mesh_elem_t elements[] = {
    ESP_BLE_MESH_ELEMENT(0, root_models, ESP_BLE_MESH_MODEL_NONE),
    ESP_BLE_MESH_ELEMENT(0, root_models1, ESP_BLE_MESH_MODEL_NONE),
    ESP_BLE_MESH_ELEMENT(0, root_models2, ESP_BLE_MESH_MODEL_NONE),
};
```

Even when i am getting those errors nRF Mesh is recognizing that node has 3 elements and i can get/set value for each element.
Yhere is one more small issue i dont understand. When with Silicon labs provisioner i have provisioned mesh node and then i turn off proxy i am getting this message constantly repeated every 10 sec:
```
W (1872995) BLE_MESH: No subnets to advertise on
W (1882995) BLE_MESH: No subnets to advertise on
```

Anyway, i just want to say one more time 🥇 Thank you espressif team.

**Comment by wnnwoo** [comment]
How about the latest  mesh status?

**Comment by jeromeDms** [comment]
> I have to say BIG thank you for releasing beta version. I had not much time to play with it so far, but i love it. I dont know yet which android provisioner i like more, with silicon labs we have options to easy switch all available server configs like proxy, relay or element type. In other hand is not recognizing more than one element per node (maybe its esp32 issue see ad1) even if is assigning all necessary unicast addresses. Second provisioner is nRF Mesh, which is not as easy to use, but is recognizing all elements in one mesh node (3 in my tests). I could not find option to turn on/off relay or proxy like in Silicon lab app, but there is more options to manage mesh node.
> 
> AD1: I dont want to open issue yet, because i dont know if this is issue or just i am initializing elements in node wrong way. This is log i am getting:
> 
> ```
> E (448) BT_LOG: btc_mesh_model_op_add, Invalid model operation min_len = 1073486928, or param
> callbak = 0x3ffc1d48, the last valid opcode = 0x0
> E (448) BT_LOG: btc_mesh_model_op_add, Invalid model operation min_len = 1073486928, or param
> callbak = 0x3ffc1d48, the last valid opcode = 0x0
> E (458) BT_LOG: btc_mesh_model_op_add, Invalid model operation min_len = 1073486928, or param
> callbak = 0x3ffc1d48, the last valid opcode = 0x0
> ```
> Code:
> 
> ```
> static esp_ble_mesh_model_t root_models[] = {
>     ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
>     ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op, &onoff_pub, &led_state[0]),
> };
> 
> static esp_ble_mesh_model_t root_models1[] = {
>     ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
>     ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op, &onoff_pub, &led_state[1]),
> };
> 
> static esp_ble_mesh_model_t root_models2[] = {
>     ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
>     ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op, &onoff_pub, &led_state[2]),
> };
> 
> static esp_ble_mesh_elem_t elements[] = {
>     ESP_BLE_MESH_ELEMENT(0, root_models, ESP_BLE_MESH_MODEL_NONE),
>     ESP_BLE_MESH_ELEMENT(0, root_models1, ESP_BLE_MESH_MODEL_NONE),
>     ESP_BLE_MESH_ELEMENT(0, root_models2, ESP_BLE_MESH_MODEL_NONE),
> };
> ```
> Even when i am getting those errors nRF Mesh is recognizing that node has 3 elements and i can get/set value for each element.
> There is one more small issue i dont understand. When with Silicon labs provisioner i have provisioned mesh node and then i turn off proxy i am getting this message constantly repeated every 10 sec:
> 
> ```
> W (1872995) BLE_MESH: No subnets to advertise on
> W (1882995) BLE_MESH: No subnets to advertise on
> ```
> Anyway, i just want to say one more time 🥇 Thank you espressif team.

Hi chegewara, I'm facing the same issue, was wondering if you find a way to declare multiple elements within a single device ? Thanks

**Comment by chegewara** [comment]
> Hi chegewara, I'm facing the same issue, was wondering if you find a way to declare multiple elements within a single device ? Thanks

In fact i did. This is code i changed:
```
static esp_ble_mesh_model_t root_models[] = {
    ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
    ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op,
    &onoff_pub, &led_state[0]),
};

static esp_ble_mesh_model_t root_models1[] = {
    // ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
    ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op,
    &onoff_pub, &led_state[1]),
};

static esp_ble_mesh_model_t root_models2[] = {
    // ESP_BLE_MESH_MODEL_CFG_SRV(&config_server),
    ESP_BLE_MESH_SIG_MODEL(ESP_BLE_MESH_MODEL_ID_GEN_ONOFF_SRV, onoff_op,
    &onoff_pub, &led_state[2]),
};

static esp_ble_mesh_elem_t elements[] = {
    ESP_BLE_MESH_ELEMENT(0, root_models, ESP_BLE_MESH_MODEL_NONE),
    ESP_BLE_MESH_ELEMENT(1, root_models1, ESP_BLE_MESH_MODEL_NONE),
    ESP_BLE_MESH_ELEMENT(2, root_models1, ESP_BLE_MESH_MODEL_NONE),
};
```
As you can see i have commented out CFG_SRV model in 2nd and 3rd elements, which is normal usage case from what i understand after studying ble mesh specs, but you can also have CFG_SRV model for additional elements if your application requires this.

**Comment by llymaximus** [comment]
@jeromeDms @chegewara 
* As per the Mesh spec says, **The Configuration Server model shall be supported by a primary element and shall not be supported by any secondary elements.**
* For multiple elements, please mask the following code in btc_ble_mesh_prov.c (line 966 ~ 971). This mask will be merged soon. 
```
if (op->min_len > BLE_MESH_MAX_DATA_SIZE || op->param_cb != 0) {
    --op;
    LOG_ERROR("%s, Invalid model operation min_len = %d, or param callbak = 0x%x, the last valid opcode = 0x%x", __func__, op->min_len, op->param_cb, op->opcode);
    return;
}
```
* Also recommend to use the Android nRF Mesh App, i found some bugs while using the iOS nRF Mesh App (version V 1.0.2) test multiple elements.

**Comment by Campou** [comment]
@jeromeDms @chegewara
* As per the Mesh spec says, **The Configuration Server model shall be supported by a primary element and shall not be supported by any secondary elements.**
* For multiple elements, please mask the following code in btc_ble_mesh_prov.c (line 966 ~ 971). This mask will be merged soon.
```
if (op->min_len > BLE_MESH_MAX_DATA_SIZE || op->param_cb != 0) {
    --op;
    LOG_ERROR("%s, Invalid model operation min_len = %d, or param callbak = 0x%x, the last valid opcode = 0x%x", __func__, op->min_len, op->param_cb, op->opcode);
    return;
}
```
* Also recommend to use the Android nRF Mesh App, i found some bugs while using the iOS nRF Mesh App (version V 1.0.2) test multiple elements.

**Comment by Alvin1Zhang** [comment]
@Campou Thanks for the updates.
